msal/broker.py

```diff
@@ -0,0 +1,235 @@
+It relies on PyMsalRuntime which is the package providing broker's functionality.
+"""
+from threading import Event
+import json
+import logging
+import time
+import uuid
+
+
+logger = logging.getLogger(__name__)
+try:
+    import pymsalruntime  # ImportError would be raised on unsupported platforms such as Windows 8
+    # Its API description is available in site-packages/pymsalruntime/PyMsalRuntime.pyi
+    pymsalruntime.register_logging_callback(lambda message, level: {  # New in pymsalruntime 0.7
+        pymsalruntime.LogLevel.TRACE: logger.debug,  # Python has no TRACE level
+        pymsalruntime.LogLevel.DEBUG: logger.debug,
+        # Let broker's excess info, warning and error logs map into default DEBUG, for now
+        #pymsalruntime.LogLevel.INFO: logger.info,
+        #pymsalruntime.LogLevel.WARNING: logger.warning,
+        #pymsalruntime.LogLevel.ERROR: logger.error,
+        pymsalruntime.LogLevel.FATAL: logger.critical,
+        }.get(level, logger.debug)(message))
+except (ImportError, AttributeError):  # AttributeError happens when a prior pymsalruntime uninstallation somehow leaved an empty folder behind
+    # PyMsalRuntime currently supports these Windows versions, listed in this MSFT internal link
+    # https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/2406/files
+    raise ImportError(  # TODO: Remove or adjust this line right before merging this PR
+        'You need to install dependency by: pip install "msal[broker] @ git+https://github.com/AzureAD/microsoft-authentication-library-for-python.git@wam"')
+    raise ImportError('You need to install dependency by: pip install "msal[broker]>=1.18,<2"')
+# Other exceptions (possibly RuntimeError) would be raised if its initialization fails
+
+
+class RedirectUriError(ValueError):
+    pass
+
+
+class _CallbackData:
+    def __init__(self):
+        self.signal = Event()
+        self.result = None
+
+    def complete(self, result):
+        self.signal.set()
+        self.result = result
+
+
+def _convert_error(error, client_id):
+    context = error.get_context()  # Available since pymsalruntime 0.0.4
+    if (
+            "AADSTS50011" in context  # In WAM, this could happen on both interactive and silent flows
+            or "AADSTS7000218" in context  # This "request body must contain ... client_secret" is just a symptom of current app has no WAM redirect_uri
+            ):
+        raise RedirectUriError(  # This would be seen by either the app developer or end user
+            "MsalRuntime won't work unless this one more redirect_uri is registered to current app: "
+            "ms-appx-web://Microsoft.AAD.BrokerPlugin/{}".format(client_id))
+        # OTOH, AAD would emit other errors when other error handling branch was hit first,
+        # so, the AADSTS50011/RedirectUriError is not guaranteed to happen.
+    return {
+        "error": "broker_error",  # Note: Broker implies your device needs to be compliant.
```

Comment:
If getting a certificate fails in Cloud Shell, MSAL also raises `broker_error`. 

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/292e28bbb5a9488de1853a985a994882781068ac/msal/cloudshell.py#L63-L68

Certainly this "broker" (WAM) is different from Cloud Shell's broker (pseudo managed identity). Expected?